app/src/main/java/com/google/android/sambadocumentsprovider/browsing/broadcast/BroadcastBrowsingProvider.java

```diff
@@ -0,0 +1,122 @@
+import com.google.android.sambadocumentsprovider.browsing.BrowsingException;
+import com.google.android.sambadocumentsprovider.browsing.NetworkBrowsingProvider;
+
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+public class BroadcastBrowsingProvider implements NetworkBrowsingProvider {
+  private static final String TAG = "BroadcastBrowsing";
+  private static final int LISTENING_TIMEOUT = 3000;
+
+  private final ExecutorService mExecutor = Executors.newFixedThreadPool(2);
+
+  @Override
+  public List<String> getServers() throws BrowsingException {
+    try {
+      DatagramSocket socket = new DatagramSocket();
+
+      mExecutor.submit(new BroadcastSender(socket));
+
+      Future<List<String>> servers = mExecutor.submit(
+              new BroadcastListener(socket, LISTENING_TIMEOUT));
+
+      return servers.get();
+    } catch(ExecutionException | InterruptedException | SocketException e){
+      Log.e(TAG, "Failed to send broadcast", e);
+
+      throw new BrowsingException("Failed to get servers via broadcast: " + e.getMessage());
+    }
+  }
+
+  private static class BroadcastSender implements Runnable {
+    private static final String TAG = "BroadcastSender";
+    private static final int NBT_PORT = 137;
+
+    private final DatagramSocket mSocket;
+
+    BroadcastSender(DatagramSocket socket) {
+      mSocket = socket;
+    }
+
+    public void run() {
+      try {
+        InetAddress local = InetAddress.getByName(BroadcastUtils.getBroadcastAddress());
+
+        byte[] data = BroadcastUtils.createPacket();
+        int dataLength = data.length;
+
+        DatagramPacket p = new DatagramPacket(data, dataLength, local, NBT_PORT);
+        mSocket.send(p);
+
+        if (BuildConfig.DEBUG) Log.d(TAG, "Broadcast package sent");
+      } catch (Exception e) {
+        Log.e(TAG, "Failed to send broadcast", e);
+      }
+    }
+  }
+
+  private static class BroadcastListener implements Callable<List<String>> {
+    private static final String TAG = "BroadcastListener";
+
+    private final DatagramSocket mSocket;
+    private final int mTimeout;
+
+    BroadcastListener(DatagramSocket socket, int timeout) {
+      mSocket = socket;
+      mTimeout = timeout;
+    }
+
+    @Override
+    public List<String> call() throws Exception {
+      if (BuildConfig.DEBUG) Log.d(TAG, "Listening for broadcast responses");
+
+      DatagramSocket socket = mSocket;
+
+      socket.setSoTimeout(mTimeout);
+
+      byte[] buf = new byte[1024];
+      DatagramPacket packet = new DatagramPacket(buf, buf.length);
+
+      while (true) {
+        try {
+          socket.receive(packet);
```

Comment:
Oh... I took it out of the loop a few changes ago and didn't test on SambaNetwork after that. Thanks, fixed that.